cql3/expr/restrictions.cc

```diff
@@ -271,5 +271,24 @@ binary_operator validate_and_prepare_new_restriction(const binary_operator& rest
     return prepared_binop;
 }
 
+void validate_single_collection_column_relation(const expression& col_expr, oper_t oper, schema_ptr schema) {
+    if (auto col_val = as_if<column_value>(&col_expr)) {
+        // Simple single collection column restriction
+        const abstract_type& lhs_col_type = col_val->col->column_specification->type->without_reversed();
+
+        if (lhs_col_type.is_collection()) {
+            validate_single_column_relation(*col_val, oper, *schema, false);
```

Comment:
If I understand this correctly, what you do here is call a pre-existing validation function that used to be called, but the new expression code forgot to call for expressions. So why do we need the `if(lhs_col_type.is_collection())` condition? Shouldn't we call the validation function unconditionally if the relation is on a column_value?

Reply:
Hi @nyh, thanks for your remarks.
The idea was to call validations from `validate_single_column_relation` before validations from `prepare_expression` called from `prepare_binary_operator`. Cause described situation occurs for collection operands only I decided to call validations here for collections only. I agree that may be a disputable solution and as a variant I had an idea to move the validation part from `validate_and_prepare_new_restriction`:

`if (auto col_val = as_if<column_value>(&prepared_binop.lhs)) {
    // Simple single column restriction
    validate_single_column_relation(*col_val, prepared_binop.op, *schema, false);
} else if (auto sub = as_if<subscript>(&prepared_binop.lhs)) {
    // Subscripted single column restriction
    const column_value& sub_col = get_subscripted_column(*sub);
    validate_single_column_relation(sub_col, prepared_binop.op, *schema, true);
}`

to `prepare_binary_operator` and call it for all types of operands. Maybe it would look better but potentially it can change validation order for other types (not collections) and as a result show incorrect messages. That's why I've chosen different way.

Reply:
hi @nyh, I've changed code according to your recommendations. Hope now it will look better for you.
However I'm not sure we should call validation for non-collection operands twice...
I tried to remove calls of `validate_single_column_relation` from `validate_and_prepare_new_restriction` and (as expected) it changed existing behaviour (many tests failed).